source/Weather.cpp

```diff
@@ -70,12 +71,13 @@ const Point &Weather::Origin() const
 
 
 // Create any environmental effects and decrease the lifetime of this weather.
-void Weather::Step(vector<Visual> &visuals)
+void Weather::Step(vector<Visual> &visuals, const Point *effectCenter)
 {
 	// Environmental effects are created by choosing a random angle and distance from
 	// their origin, then creating the effect there.
 	double minRange = hazard->MinRange();
-	double maxRange = hazard->MaxRange();
+	// If it is systemwide, only draw around the flagship.
+	double maxRange = hazard->SystemWide() ? 5000. : hazard->MaxRange();
```

Comment:
If some hazard is really system-wide, then we should be able to implement it in a way that there is no range at all needed.

Reply:
And I guess we should have some PRs to optimize not drawing the ships (and their effects) that are "outside our view". But that sounds like a lot of work which is likely best kept for a separate PR.

Reply:
> And I guess we should have some PRs to optimize not drawing the ships (and their effects) that are "outside our view"

If I'm not missing anything, this is already the case. Bodies that are outside of the screen are culled and discarded.

Reply:
Then what was causing the lag some where having with big range hazards and a lot of ships was having to check if it applies to all ships, right?
How could hazard be made to work without range though?

Reply:
> If some hazard is really system-wide, then we should be able to implement it in a way that there is no range at all needed.

This is all for optimization's sake, given effects can make the game lag system-wide hazards are not generated for the entire system. Instead they are drawn around the flagship (in big enough of a field so that the player wont notice anything), and are applied to every single ship in the system without further checks.

Reply:
> This is all for optimization's sake,

Yes, I understood that. I would actually prefer another optimization where we have a system-wide effect, and where we optimize that effect to only draw the effects visible in the active view/screen. If we do that, then we optimize even more by not drawing any non-visible effects. (And people with large screens also don't get strange artifacts on the side.)

Reply:
But even in the case you are describing we would still generate some things to not draw them later?

Reply:
> Yes, I understood that. I would actually prefer another optimization where we have a real system-wide effect, and where we then optimize that effect to only draw the effects visible in the active view/screen. If we do that, then we optimize even more by not drawing any non-visible effects. 

You're taking about not drawing things, I'm taking about not generating them in the first place, so I optimize more ^^

> (And people with large screens also don't get strange artifacts on the side.)

Now I based it off the size of the screen, so that won't happen.
After all I did not take zoom into account and just did the drawing assuming it was largest (caus otherwise zooming out as the hazard ends means you notice it wasn't actually visually system-wide)

Reply:
The best optimization is indeed to only _generate_ the visual effects that are visible on the screen. If we can restrict the generation to only the visible screen, then in theory we would have the best performance possible.